crypto/dh/dh_check.c

```diff
@@ -18,6 +18,18 @@
  * p is odd
  * 1 < g < p - 1
  */
+int DH_check_params_ex(const DH *dh)
+{
+    int errflags = 0;
+    int ok = DH_check_params(dh, &errflags);
+
+    if ((errflags & DH_CHECK_P_NOT_PRIME) != 0)
+        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_CHECK_P_NOT_PRIME);
+    if ((errflags & DH_NOT_SUITABLE_GENERATOR) != 0)
+        DHerr(DH_F_DH_CHECK_PARAMS_EX, DH_R_NOT_SUITABLE_GENERATOR);
+
+    return ok;
```

Comment:
OK, so here's where things are getting misunderstood.  `DH_check_params` returns 1 even though if might find things wrong with the params (which is correct insofar that the checking operation successfully found the state of the key)...  But since `DH_check_params_ex` generates *errors* if there's something wrong with the key, returning 1 is just plain wrong.

This return statement should be:

``` C
    return errflags == 0;
```

Reply:
Ahh, the original DH check functions are very inconsistent with other 'check' functions and I didn't look into them very carefully...

Reply:
Fixed